### ISSUE TITLE
When a help message is associated with a warning message, the two message you should be printed together and the source code printed only once.

### ISSUE THREAD

**Issue by dzamlo** [rust-lang/rust#21938]
For exemple, for the following code:
```rust
extern crate libc;

fn main() {
}
```

The output of rustc is:
```
bar.rs:1:1: 1:19 warning: use of unstable library feature 'libc'
bar.rs:1 extern crate libc;
         ^~~~~~~~~~~~~~~~~~
bar.rs:1:1: 1:19 help: add #![feature(libc)] to the crate attributes to silence this warning
bar.rs:1 extern crate libc;
         ^~~~~~~~~~~~~~~~~~
```

It should be:
```
bar.rs:1:1: 1:19 warning: use of unstable library feature 'libc'
bar.rs:1:1: 1:19 help: add #![feature(libc)] to the crate attributes to silence this warning
bar.rs:1 extern crate libc;
         ^~~~~~~~~~~~~~~~~~
```

I don't know if there is any help message associated with an error message, but if this the case, the same issue apply.

**Comment by caipre** [comment]
Looks like the referenced merge addresses error + help combos (albeit not in the order described above):
```console
$ echo 'mod foo;' > test.rs
$ rustc ./test.rs
test.rs:1:5: 1:8 error: file not found for module `foo`
test.rs:1 mod foo;
              ^~~
test.rs:1:5: 1:8 help: name the file either foo.rs or foo/mod.rs inside the directory ""
```

But this still happens for error + warning combos:
```rust
use std::vec::Drain;

fn main () {
}
```

```console
$ rustc ./test.rs
./test.rs:1:5: 1:20 error: use of unstable library feature 'collections': recently added as part of collections reform 2
./test.rs:1 use std::vec::Drain;
                ^~~~~~~~~~~~~~~
./test.rs:1:5: 1:20 warning: unused import, #[warn(unused_imports)] on by default
./test.rs:1 use std::vec::Drain;
                ^~~~~~~~~~~~~~~
error: aborting due to previous error
```

Though this may be deliberate since the error and warning are about different issues.

**Comment by carols10cents** [comment]
> (albeit not in the order described above):

So since [helps look like they're a subdiagnostic of errors](https://github.com/rust-lang/rust/blob/142214d1f2232a4e88ff7bd99951b01f36052c61/src/libsyntax/errors/mod.rs#L252), and [errors get output before their subdiagnostics](https://github.com/rust-lang/rust/blob/142214d1f2232a4e88ff7bd99951b01f36052c61/src/libsyntax/errors/emitter.rs#L32-L33), the ordering looks intentional.

With the error+warning case, I think those are actually two separate diagnostics, that is, the warning is not the subdiagnostic of the error, so they shouldn't be combined.

**Sooo I think this issue might be done?** If I'm understanding the intentionality of the current error display?

**Comment by brson** [comment]
Warnings have undergone a lot of churn lately, and some of them have been grouped together in logical ways. @jonathandturner @nikomatsakis @nrc @pnkfelix is this issue still relevant?

**Comment by pnkfelix** [comment]
I think we can close this. The main redundancy has been addressed,  and I think the current handling of help messages works at least for the case described

**Comment by nikomatsakis** [comment]
Agreed.
